JodaCompatibleZonedDateTime in Elasticsearch 6.5.x breaks scripts with "The datetime zone id 'Z' is not recognised"

**Elasticsearch version** 6.5.1

**Plugins installed**: []

**JVM version** JVM 1.8.0_191

**OS version** Windows 10

**Description of the problem including expected versus actual behavior**:

We use scripting with date functions in some code. When we try to upgrade to Elasticsearch 6.5.1, one of our unit tests fails with the following exception:

```
Caused by: java.lang.IllegalArgumentException: The datetime zone id 'Z' is not recognised
	at org.joda.time.DateTimeZone.forID(DateTimeZone.java:234)
	at org.elasticsearch.script.JodaCompatibleZonedDateTime.toString(JodaCompatibleZonedDateTime.java:405)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.elasticsearch.painless.DefBootstrap$PIC.fallback(DefBootstrap.java:216)
	... 32 more
```

**Steps to reproduce**:

The following also reproduces this for me, it seems Joda does not know about the special "Z" timezone identifier, but Java Time uses it:

```
    @Test
    public void testJodaCompatibleZonedDateTime() {
        // check for an issue that we see with 6.5
        JodaCompatibleZonedDateTime dt = new JodaCompatibleZonedDateTime(Instant.EPOCH, ZoneOffset.ofTotalSeconds(0));
        assertNotNull(dt.toString("yyyy-MM-dd hh:mm", Locale.ROOT));
    }
```

It seems this was introduced with the Java-time-compatibility layer in commit 758a6f52e299891ac2d1c2d9128af0fe0c21baac and 3046656ab11b0018729dc158d764b8f6246ec92d, PR #33486


A more elaborate integration test that works at least in Elasticsearch 6.2.4, but fails with Elasticsearch 6.5.1:

 ```
   @Test
    public void testScriptDateTime() {
        index("idx", "_doc", "id1", "{ \"endTime\": 0}");
        refresh();

        final SearchRequestBuilder searchRequestBuilder = client().prepareSearch("*").
                setSize(100).
                addAggregation(AggregationBuilders.terms("script").
                        script(new Script(ScriptType.INLINE, "painless",
                                "doc['endTime'].date.toString('yyyy-MM-dd hh:mm', Locale.ROOT)", Collections.emptyMap())));
        final SearchResponse searchResponse = searchRequestBuilder.execute().actionGet();
        assertNotNull(searchResponse);
        final Terms terms = searchResponse.getAggregations().get("script");
        assertNotNull(terms.getBuckets());
        assertEquals("1970-01-01 12:00", terms.getBuckets().get(0).getKeyAsString());
        assertEquals(1, searchResponse.getHits().getTotalHits());
    }
```

--- a/server/src/main/java/org/elasticsearch/script/JodaCompatibleZonedDateTime.java
+++ b/server/src/main/java/org/elasticsearch/script/JodaCompatibleZonedDateTime.java
@@ -22,8 +22,8 @@ package org.elasticsearch.script;
 import org.apache.logging.log4j.LogManager;
 import org.elasticsearch.common.SuppressForbidden;
 import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.time.DateUtils;
 import org.joda.time.DateTime;
-import org.joda.time.DateTimeZone;
 
 import java.security.AccessController;
 import java.security.PrivilegedAction;
@@ -391,14 +391,14 @@ public class JodaCompatibleZonedDateTime {
     public String toString(String format) {
         logDeprecatedMethod("toString(String)", "a DateTimeFormatter");
         // TODO: replace with bwc formatter
-        return new DateTime(dt.toInstant().toEpochMilli(), DateTimeZone.forID(dt.getZone().getId())).toString(format);
+        return new DateTime(dt.toInstant().toEpochMilli(), DateUtils.zoneIdToDateTimeZone(dt.getZone())).toString(format);
     }
 
     @Deprecated
     public String toString(String format, Locale locale) {
         logDeprecatedMethod("toString(String,Locale)", "a DateTimeFormatter");
         // TODO: replace with bwc formatter
-        return new DateTime(dt.toInstant().toEpochMilli(), DateTimeZone.forID(dt.getZone().getId())).toString(format, locale);
+        return new DateTime(dt.toInstant().toEpochMilli(), DateUtils.zoneIdToDateTimeZone(dt.getZone())).toString(format, locale);
     }
 
     public DayOfWeek getDayOfWeekEnum() {

--- a/server/src/test/java/org/elasticsearch/script/JodaCompatibleZonedDateTimeTests.java
+++ b/server/src/test/java/org/elasticsearch/script/JodaCompatibleZonedDateTimeTests.java
@@ -237,4 +237,14 @@ public class JodaCompatibleZonedDateTimeTests extends ESTestCase {
     public void testDayOfWeekEnum() {
         assertThat(javaTime.getDayOfWeekEnum(), equalTo(DayOfWeek.of(jodaTime.getDayOfWeek())));
     }
+
+    public void testToStringWithLocaleAndZeroOffset() {
+        JodaCompatibleZonedDateTime dt = new JodaCompatibleZonedDateTime(Instant.EPOCH, ZoneOffset.ofTotalSeconds(0));
+        assertMethodDeprecation(() -> dt.toString("yyyy-MM-dd hh:mm", Locale.ROOT), "toString(String,Locale)", "a DateTimeFormatter");
+    }
+
+    public void testToStringAndZeroOffset() {
+        JodaCompatibleZonedDateTime dt = new JodaCompatibleZonedDateTime(Instant.EPOCH, ZoneOffset.ofTotalSeconds(0));
+        assertMethodDeprecation(() -> dt.toString("yyyy-MM-dd hh:mm"), "toString(String)", "a DateTimeFormatter");
+    }
 }

